hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java

```diff
@@ -108,567 +75,151 @@
-
-            //Remove existing deps
-            pythonDepsFacade.removePythonDepsForProject(projectId);
-
-            //Insert all deps in current listing
-            pythonDepsFacade.addPythonDepsForProject(projectId, pythonDeps);
-          }
-
-        }
-
-        //an upgrade results in an unknown version installed, query local conda env to figure it out
-        if (command.getOp().equals(CondaOp.UPGRADE)) {
-
-          command.setVersion(getLocalLibraryVersion(command.getLib(), command.getVersion(), projName));
-
-          if (settings.getHopsworksIp().equals(command.getHostId().getHostIp())) {
-            Project projectId = command.getProjectId();
-
-            Collection<PythonDep> pythonDeps = projectId.getPythonDepCollection();
-            for (PythonDep pythonDep : pythonDeps) {
-              if (pythonDep.getDependency().equals(command.getLib()) && pythonDep.getVersion().equals(command.
-                  getVersion())) {
-
-                String localVersion = getLocalLibraryVersion(command.getLib(), command.getVersion(), projName);
-
-                if (!localVersion.equals(command.getVersion())) {
-                  Collection<PythonDep> deps = projectId.getPythonDepCollection();
-
-                  for (PythonDep dep : deps) {
-                    if (dep.getDependency().equals(command.getLib())) {
-                      PythonDep newDep = pythonDepsFacade.getDep(dep.getRepoUrl(), dep.getMachineType(),
-                          command.getInstallType(), command.getLib(), localVersion, true, false);
-                      deps.remove(dep);
-                      deps.add(newDep);
-                      projFacade.update(projectId);
-                      break;
-                    }
-                  }
-                }
-              }
-            }
-          }
-        }
-      }
-    }
-  }
-
-  private void emailAlert(String subject, String body) {
-    try {
-      emailBean.sendEmails(settings.getAlertEmailAddrs(), subject, body);
-    } catch (MessagingException ex) {
-      logger.log(Level.SEVERE, ex.getMessage());
+  public Response handle(AgentView request,
+      @QueryParam("action")
+      @DefaultValue("NONE") AgentAction action)
+    throws AppException {
+    switch (action) {
+      case PING:
+        logger.log(Level.FINE, "AgentResource: Ping");
+        handlePing();
+        break;
+      case HEARTBEAT:
+        logger.log(Level.FINE, "AgentResource: Heartbeat");
+        final HeartbeatReplyDTO commands = handleHeartbeat(request);
+        
+        final AgentView agentHBReply = heartbeatReplyToAgentView(commands);
+        
+        return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK)
+            .entity(agentHBReply).build();
+      case REGISTER:
+        logger.log(Level.FINE, "AgentResource: Register");
+        final String hadoopHome = handleRegister(request);
+        final AgentView agentRegReply = new AgentView();
```

Comment:
What's the purpose of all these `finals`?

Reply:
The heartbeat passes through several layers and I want to be sure it is not re-assigned accidentally, but mainly to emphasize it is and semantically *should* remain immutable